README.rst
========
Overview
========

.. start-badges

.. list-table::
    :stub-columns: 1

    * - docs
      - |docs|
    * - tests
      - | |travis| |appveyor| |requires|
        | |codecov|
    * - package
      - |commits-since|
.. |docs| image:: https://readthedocs.org/projects/tsp-hiram/badge/?style=flat
    :target: https://readthedocs.org/projects/tsp-hiram
    :alt: Documentation Status

.. |travis| image:: https://api.travis-ci.org/hiramf/tsp-hiram.svg?branch=master
    :alt: Travis-CI Build Status
    :target: https://travis-ci.org/hiramf/tsp-hiram

.. |appveyor| image:: https://ci.appveyor.com/api/projects/status/github/hiramf/tsp-hiram?branch=master&svg=true
    :alt: AppVeyor Build Status
    :target: https://ci.appveyor.com/project/hiramf/tsp-hiram

.. |requires| image:: https://requires.io/github/hiramf/tsp-hiram/requirements.svg?branch=master
    :alt: Requirements Status
    :target: https://requires.io/github/hiramf/tsp-hiram/requirements/?branch=master

.. |codecov| image:: https://codecov.io/gh/hiramf/tsp-hiram/branch/master/graphs/badge.svg?branch=master
    :alt: Coverage Status
    :target: https://codecov.io/github/hiramf/tsp-hiram

.. |commits-since| image:: https://img.shields.io/github/commits-since/hiramf/tsp-hiram/v0.2.1.svg
    :alt: Commits since latest release
    :target: https://github.com/hiramf/tsp-hiram/compare/v0.2.1...master



.. end-badges

Hiram Foster's solution to the Traveling Salesman coding challenge.

There were two problems to solve:

#. Given an input of list of co-ordinates find the path which is the shortest distance which touches ALL these points
#. Given an input of list of co-ordinates and fixed distance identify the maximum number of points which can be touched

Solution Artifacts required:

+ Working Code
+ Unit Test Cases (can be run with Tox/pytest)
+ Exception and Edge Cases: Documented in the log and accounted for in the code
+ Running `log <https://tsp-hiram.readthedocs.io/en/latest/log.html>`_ of the steps/process and everything which went on to solve this problem


Installation
============

You can install the in-development version with::

    pip install https://github.com/hiramf/tsp-hiram/archive/master.zip


Documentation
=============

This is a module than can be installed like any other normal Python package and imported into scripts, Jupyter, etc.

To use the CLI, run:

.. code-block:: python

   tsp-hiram filename.csv --max 100

Where ``filename.csv`` is a two column csv file of coordinates with a header and ``--max 100`` is an optional argument to set the maximum distance of the solution.



https://tsp-hiram.readthedocs.io/


Development
===========

To run the all tests run::

    tox

Note, to combine the coverage data from all the tox environments run:

.. list-table::
    :widths: 10 90
    :stub-columns: 1

    - - Windows
      - ::

            set PYTEST_ADDOPTS=--cov-append
            tox

    - - Other
      - ::

            PYTEST_ADDOPTS=--cov-append tox


CHANGELOG.rst

Changelog
=========

0.0.0 (2020-05-06)
------------------

* First release on PyPI.


tests/test_cli.py
import csv
import sys

import pytest

from tsp_hiram.cli import main


@pytest.fixture
def coordinates_list():
    data = [
        [8, 41],
        [220, 125],
        [56, 89],
        [196, 145],
        [196, 49],
        [64, 21],
        [124, 117],
        [212, 65],
        [24, 17],
        [24, 25],
        [172, 117],
        [260, 109],
        [252, 21],
        [172, 145],
        [40, 137],
    ]

    return data


@pytest.mark.parametrize("max_distance,expected", [(None, 750), (128, 126)])
def test_main(monkeypatch, tmp_path, coordinates_list, max_distance, expected):
    """Tests the CLI with and without max distance parameter
    """
    # Create a temporary file called "coordinates.csv"
    d = tmp_path / "sub"
    d.mkdir()
    p = d / "coordinates.csv"
    filepath = str(p)
    with open(p, 'w', newline='') as csvfile:
        writer = csv.writer(csvfile, delimiter=",")
        writer.writerow(['x', 'y'])
        for row in coordinates_list:
            writer.writerow(row)

    # Test the CLI with different args
    with monkeypatch.context() as m:
        args = ['main', filepath]
        if max_distance is not None:
            args += ['--max', str(max_distance)]
        m.setattr(sys, 'argv', args)
        route, distance = main()
        assert distance == expected


tests/test_tsp.py
import math
import random

import numpy as np
import pytest

from tsp_hiram import tsp


@pytest.fixture
def coordinates_google(scope="module"):
    data = [
        (288, 149), (288, 129), (270, 133), (256, 141), (256, 157), (246, 157),
        (236, 169), (228, 169), (228, 161), (220, 169), (212, 169), (204, 169),
        (196, 169), (188, 169), (196, 161), (188, 145), (172, 145), (164, 145),
        (156, 145), (148, 145), (140, 145), (148, 169), (164, 169), (172, 169),
        (156, 169), (140, 169), (132, 169), (124, 169), (116, 161), (104, 153),
        (104, 161), (104, 169), (90, 165), (80, 157), (64, 157), (64, 165),
        (56, 169), (56, 161), (56, 153), (56, 145), (56, 137), (56, 129),
        (56, 121), (40, 121), (40, 129), (40, 137), (40, 145), (40, 153),
        (40, 161), (40, 169), (32, 169), (32, 161), (32, 153), (32, 145),
        (32, 137), (32, 129), (32, 121), (32, 113), (40, 113), (56, 113),
        (56, 105), (48, 99), (40, 99), (32, 97), (32, 89), (24, 89),
        (16, 97), (16, 109), (8, 109), (8, 97), (8, 89), (8, 81),
        (8, 73), (8, 65), (8, 57), (16, 57), (8, 49), (8, 41),
        (24, 45), (32, 41), (32, 49), (32, 57), (32, 65), (32, 73),
        (32, 81), (40, 83), (40, 73), (40, 63), (40, 51), (44, 43),
        (44, 35), (44, 27), (32, 25), (24, 25), (16, 25), (16, 17),
        (24, 17), (32, 17), (44, 11), (56, 9), (56, 17), (56, 25),
        (56, 33), (56, 41), (64, 41), (72, 41), (72, 49), (56, 49),
        (48, 51), (56, 57), (56, 65), (48, 63), (48, 73), (56, 73),
        (56, 81), (48, 83), (56, 89), (56, 97), (104, 97), (104, 105),
        (104, 113), (104, 121), (104, 129), (104, 137), (104, 145), (116, 145),
        (124, 145), (132, 145), (132, 137), (140, 137), (148, 137), (156, 137),
        (164, 137), (172, 125), (172, 117), (172, 109), (172, 101), (172, 93),
        (172, 85), (180, 85), (180, 77), (180, 69), (180, 61), (180, 53),
        (172, 53), (172, 61), (172, 69), (172, 77), (164, 81), (148, 85),
        (124, 85), (124, 93), (124, 109), (124, 125), (124, 117), (124, 101),
        (104, 89), (104, 81), (104, 73), (104, 65), (104, 49), (104, 41),
        (104, 33), (104, 25), (104, 17), (92, 9), (80, 9), (72, 9),
        (64, 21), (72, 25), (80, 25), (80, 25), (80, 41), (88, 49),
        (104, 57), (124, 69), (124, 77), (132, 81), (140, 65), (132, 61),
        (124, 61), (124, 53), (124, 45), (124, 37), (124, 29), (132, 21),
        (124, 21), (120, 9), (128, 9), (136, 9), (148, 9), (162, 9),
        (156, 25), (172, 21), (180, 21), (180, 29), (172, 29), (172, 37),
        (172, 45), (180, 45), (180, 37), (188, 41), (196, 49), (204, 57),
        (212, 65), (220, 73), (228, 69), (228, 77), (236, 77), (236, 69),
        (236, 61), (228, 61), (228, 53), (236, 53), (236, 45), (228, 45),
        (228, 37), (236, 37), (236, 29), (228, 29), (228, 21), (236, 21),
        (252, 21), (260, 29), (260, 37), (260, 45), (260, 53), (260, 61),
        (260, 69), (260, 77), (276, 77), (276, 69), (276, 61), (276, 53),
        (284, 53), (284, 61), (284, 69), (284, 77), (284, 85), (284, 93),
        (284, 101), (288, 109), (280, 109), (276, 101), (276, 93), (276, 85),
        (268, 97), (260, 109), (252, 101), (260, 93), (260, 85), (236, 85),
        (228, 85), (228, 93), (236, 93), (236, 101), (228, 101), (228, 109),
        (228, 117), (228, 125), (220, 125), (212, 117), (204, 109), (196, 101),
        (188, 93), (180, 93), (180, 101), (180, 109), (180, 117), (180, 125),
        (196, 145), (204, 145), (212, 145), (220, 145), (228, 145), (236, 145),
        (246, 141), (252, 125), (260, 129), (280, 133)
    ]

    return data


@pytest.fixture
def distance_matrix_mip(scope="module"):
    distances = [
        [0, 83, 81, 113, 52, 42, 73, 44, 23, 91, 105, 90, 124, 57],
        [83, 0, 161, 160, 39, 89, 151, 110, 90, 99, 177, 143, 193, 100],
        [81, 161, 0, 90, 125, 82, 13, 57, 71, 123, 38, 72, 59, 82],
        [113, 160, 90, 0, 123, 77, 81, 71, 91, 72, 64, 24, 62, 63],
        [52, 39, 125, 123, 0, 51, 114, 72, 54, 69, 139, 105, 155, 62],
        [42, 89, 82, 77, 51, 0, 70, 25, 22, 52, 90, 56, 105, 16],
        [73, 151, 13, 81, 114, 70, 0, 45, 61, 111, 36, 61, 57, 70],
        [44, 110, 57, 71, 72, 25, 45, 0, 23, 71, 67, 48, 85, 29],
        [23, 90, 71, 91, 54, 22, 61, 23, 0, 74, 89, 69, 107, 36],
        [91, 99, 123, 72, 69, 52, 111, 71, 74, 0, 117, 65, 125, 43],
        [105, 177, 38, 64, 139, 90, 36, 67, 89, 117, 0, 54, 22, 84],
        [90, 143, 72, 24, 105, 56, 61, 48, 69, 65, 54, 0, 60, 44],
        [124, 193, 59, 62, 155, 105, 57, 85, 107, 125, 22, 60, 0, 97],
        [57, 100, 82, 63, 62, 16, 70, 29, 36, 43, 84, 44, 97, 0]
    ]

    return distances


def test_distance_matrix():
    """Test that computing distances from coordinates gives expected output.
    """
    coordinates = [(288, 149), (288, 129), (270, 133)]
    edm = tsp.compute_euclidean_distance_matrix(coordinates)
    assert edm == [
        [0, 20, 24],
        [20, 0, 18],
        [24, 18, 0],
    ]


def test_branch_and_cut(distance_matrix_mip):
    """Test the correct solution is returned from the branch and cut algorithm according to distance of solution.
    """
    route_matrix, distance = tsp.branch_and_cut(distance_matrix_mip)
    assert int(distance) == 547


def test_nearest_neighbor_shortest_full_route(distance_matrix_mip):
    """Tests that the nearest neighbor algorithm finds the shortest possible route by iterating over each node as a starting point.
    Because max_distance is not set, it finds a full-loop.
    """
    best_distance = math.inf
    best_route = None

    for vertex in range(len(distance_matrix_mip)):
        route_matrix, distance = tsp.nearest_neighbor_path(distance_matrix_mip, start=vertex)

        # Make sure the route touches all the nodes
        assert np.sum(route_matrix) == len(distance_matrix_mip)

        if distance < best_distance:
            best_distance = distance
            best_route = route_matrix

    assert tsp.get_edges_from_route_matrix(best_route) == [
        (0, 8), (8, 5), (5, 13), (13, 7), (7, 6), (6, 2), (2, 10), (10, 12), (12, 11), (11, 3), (3, 9), (9, 4), (4, 1), (1, 0)
    ]


@pytest.mark.parametrize("max_distance,expected", [(0, 0), (150, 6), (200, 7), (10000, 14)])
def test_nearest_neighbor_most_nodes_open(distance_matrix_mip, max_distance, expected):
    """Test the goal of finding the most number of nodes given a fixed distance.
    """
    most_nodes = 0
    for vertex in range(len(distance_matrix_mip)):
        route_matrix, _ = tsp.nearest_neighbor_path(distance_matrix_mip, closed=False, max_distance=max_distance, start=vertex)
        route = tsp.get_edges_from_route_matrix(route_matrix)

        if len(route) > most_nodes:
            most_nodes = len(route)

    assert most_nodes == expected


@pytest.mark.parametrize("max_distance,expected", [(0, 0), (150, 5), (200, 6), (10000, 14)])
def test_nearest_neighbor_most_nodes_closed(distance_matrix_mip, max_distance, expected):
    """Test the goal of finding the most number of nodes given a fixed distance.
    """
    most_nodes = 0
    for vertex in range(len(distance_matrix_mip)):
        route_matrix, _ = tsp.nearest_neighbor_path(distance_matrix_mip, closed=True, max_distance=max_distance, start=vertex)
        route = tsp.get_edges_from_route_matrix(route_matrix)

        if len(route) > most_nodes:
            most_nodes = len(route)

    assert most_nodes == expected


@pytest.mark.parametrize("closed", [True, False])
@pytest.mark.parametrize("max_distance", [0, 50, 200, 10000])
def test_optimize(coordinates_google, closed, max_distance):
    """Test limiting a solution to a fixed distance for both open and closed loops for various distances.
    Makes sure none of the nodes are repeated in a route.

    :param closed: Closed or open loop solution
    :type closed: bool
    :param max_distance: Maximum distance for the solution
    :type max_distance: int
    """
    coordinates = random.sample(coordinates_google, 10)
    n = len(coordinates)
    for i in range(n):
        route, distance = tsp.optimize(coordinates, starting_node=i, max_distance=max_distance, closed=closed)
        route_0 = [edge[0] for edge in route]  # depart
        route_1 = [edge[1] for edge in route]  # arrive
        assert len(route_0) == len(set(route_0))
        assert len(route_1) == len(set(route_1))
        assert distance <= max_distance


tests/test_tsp_hiram.py

from tsp_hiram.cli import main


def test_main():
    main([])


src/tsp_hiram/cli.py
import argparse
import csv

from tsp_hiram import tsp

parser = argparse.ArgumentParser(description='Solve Traveling Salesman Problem given a list of coordinates')
parser.add_argument('filename', metavar='FILENAME', type=str, help="File containing the coordinates.")
parser.add_argument('--max', type=int, default=None, help='Max distance of the route')
parser.add_argument('--closed', type=bool, default=False, help='Whether a constrained solutions should be an open or closed loop')


def main(args=None):
    args = parser.parse_args(args=args)
    with open(args.filename, newline='') as csvfile:
        reader = csv.reader(csvfile, delimiter=',')
        # TODO: More robust handling of csvreader. Currently just drops first row as header and makes a lot of assumptions
        coordinates = [(int(x), int(y)) for [x, y] in [row for row in reader][1:]]
        route, distance = tsp.optimize(coordinates, max_distance=args.max, closed=args.closed)

        if args.max is not None:
            print(f'{len(route)} nodes could be touched with max distance of {args.max}')
        else:
            print(f'Solution with distance of {distance} found: {route}')

        return route, distance


src/tsp_hiram/tsp.py
import logging
import math
import random
from typing import List, Tuple

import numpy as np
from mip import BINARY, Model, OptimizationStatus, minimize, xsum

logging.basicConfig(format='%(process)d-%(levelname)s-%(message)s', level=logging.DEBUG)

Coordinate = Tuple[int, int]
CoordinatesVector = List[Coordinate]
Matrix = List[List[int]]


def compute_euclidean_distance_matrix(coordinates: CoordinatesVector) -> Matrix:
    """Creates a matrix of euclidian distances between the coordinates.
    Modified from https://developers.google.com/optimization/routing/tsp#euclid_distance

    :param coordinates: A list of 2D coordinates
    :type coordinates: CoordinatesVector
    :return: 2D matrix of euclidean distance between each coordinate
    :rtype: Matrix
    """
    V = range(len(coordinates))
    distances = [[0 for j in V] for i in V]
    for from_counter, from_node in enumerate(coordinates):
        for to_counter, to_node in enumerate(coordinates):
            if from_counter == to_counter:
                pass
            else:
                # Euclidean distance
                distances[from_counter][to_counter] = (int(
                    math.hypot((from_node[0] - to_node[0]),
                               (from_node[1] - to_node[1]))))
    return distances


def get_edges_from_route_matrix(route_matrix: Matrix) -> List[Tuple]:
    """Returns a list of the edges used in a route according to the route matrix

    :param route_matrix: A matrix indicating which edges contain the optimal route
    :type route_matrix: Matrix
    :return: The row and column for the edge in the matrix
    :rtype: Tuple
    :yield: List of tuples for each edge connecting two nodes
    :rtype: List[Tuple]
    """

    def get_first_row(route_matrix):
        for row in range(len(route_matrix)):
            nodes_in_row = sum(route_matrix[row])
            if nodes_in_row == 1:
                return row
            elif nodes_in_row == 0:
                continue
            else:
                raise ValueError(f'Invalid number of nodes in row: {nodes_in_row}')

    def get_next_node_from_row(i, route_matrix):
        for j in range(len(route_matrix)):
            if route_matrix[i][j] == 1:
                return (i, j)
        raise ValueError(f"Node {i} is not connected to another node.")

    edges = []
    route_length = np.sum(route_matrix)
    row = get_first_row(route_matrix)

    while len(edges) < route_length:
        try:
            to_node = get_next_node_from_row(row, route_matrix)
            row = to_node[1]
            edges.append(to_node)
        except ValueError:
            logging.info('End of open route found.')
            # transpose the matrix
            route_matrix = [[route_matrix[j][i] for j in range(len(route_matrix))] for i in range(len(route_matrix))]
            # reverse the edges
            edges = [(edges[-1][1], edges[-1][0])]
            row = edges[0][1]
    return edges


def branch_and_cut(distance_matrix: Matrix, max_seconds=20) -> Tuple[Matrix, int]:
    """Solves a distance matrix for the shortest path

    :param distance_matrix: Distance maxtrix such as returned from `compute_euclidean_distance_matrix`
    :type distance_matrix: Matrix
    :param max_seconds: The max execution time in seconds
    :type max_seconds: int
    :return:  A matrix indicating which edges contain the optimal route, the distance of the route
    :rtype: Tuple[Matrix, int]
    """

    n = len(distance_matrix)
    model = Model()

    # binary variables indicating if arc (i,j) is used on the route or not
    x = [[model.add_var(name=f'x({i},{j})', var_type=BINARY) for j in range(n)] for i in range(n)]

    # continuous variable to prevent subtours: each city will have a
    # different sequential id in the planned route except the first one
    y = [model.add_var(name=f'y({i})') for i in range(n)]

    # objective function: minimize the distance
    model.objective = minimize(xsum(distance_matrix[i][j]*x[i][j] for i in range(n) for j in range(n)))

    # constraint : enter each city coming from another city
    for i in range(n):
        model += xsum(x[j][i] for j in range(n) if j != i) == 1

    # constraint : leave each city coming from another city
    for i in range(n):
        model += xsum(x[i][j] for j in range(n) if j != i) == 1

    # subtour elimination
    for i in range(1, n):
        for j in [x for x in range(1, n) if x != i]:
            model += y[i] - (n+1)*x[i][j] >= y[j]-n, 'noSub({},{})'.format(i, j)

    # Use nearest neighbors to find an initial feasible solution
    feasible_rm, feasible_distance = nearest_neighbor_path(distance_matrix, closed=True)
    model.start = [(x[i][j], float(feasible_rm[i][j])) for i in range(n) for j in range(n)]

    # optimizing
    model.optimize(max_seconds=max_seconds)
    if model.status in [OptimizationStatus.OPTIMAL, OptimizationStatus.FEASIBLE]:
        logging.info(f'Best route found has length {model.objective_value}.')

        route_matrix = [[int(x[i][j].x) for j in range(n)] for i in range(n)]
        return route_matrix, model.objective_value

    else:
        logging.info('No solution found.')
        route_matrix = [[0 for j in range(n)] for i in range(n)]
        return route_matrix, 0


def nearest_neighbor_path(distance_matrix: Matrix, closed=False, start: int = None, max_distance: int = None) -> Tuple[Matrix, int]:
    """Simple nearest neighbor algorithm for finding a feasible path for the Traveling Salesman Problem.

    :param distance_matrix: matrix for the cost of each edge
    :type distance_matrix: List
    :param closed: If a max_distance is set, whether the solution should be an open or closed loop, defaults to Open loop (False)
    :type closed: bool, optional
    :param start: The node to start to search for a solution, defaults to a randomly chosen node
    :type start: int, optional
    :param max_distance: If None, find full path. Otherwise, constrain cost of the path to be >= max_distance, defaults to None
    :type max_distance: int, optional
    :return:  A matrix indicating which edges contain the optimal route, the distance of the route
    :rtype: Tuple[Matrix, int]
    """

    n = len(distance_matrix)

    if start is None:
        start = random.randrange(0, len(distance_matrix))
        logging.info(f'Choosing random starting node: {start}.')

    if max_distance is None:
        max_distance = math.inf

    # Initialize all vertices as unvisited except for self
    visited = [False for v in range(n)]
    visited[start] = True

    route = [start]

    route_matrix = [[0 for j in range(n)] for i in range(n)]

    from_node = start
    distance = 0
    while distance <= max_distance:

        # Have we visited all the nodes?
        if sum(visited) == len(visited):
            # then let's go home
            distance += distance_matrix[from_node][start]
            route_matrix[from_node][start] = 1
            break

        # Find out the shortest edge connecting the current vertex and an unvisited vertex
        shortest_edge = min([distance_matrix[from_node][to_node] for to_node in range(n) if visited[to_node] is False])

        next_nodes = [node for node in range(n) if distance_matrix[from_node][node] == shortest_edge]
        next_nodes = [node for node in next_nodes if visited[node] is False]
        to_node = next_nodes[0]
        go_home_cost = distance_matrix[to_node][start] + shortest_edge

        if len(next_nodes) > 1:
            # more than one nearest neighbor, choosing randomly
            logging.info('more than one neighor found.')
            to_node = random.choice(next_nodes)

        # do we have enough to get home if we go the next edge?
        if closed & (distance + go_home_cost >= max_distance):
            # then let's go home
            distance += distance_matrix[from_node][start]
            if from_node != start:
                route_matrix[from_node][start] = 1
            break

        elif distance + shortest_edge >= max_distance:
            break

        else:
            # then keep going!
            distance += shortest_edge
            visited[to_node] = True
            route_matrix[from_node][to_node] = 1
            route.append(to_node)
            from_node = to_node

    logging.info(f'{sum(visited)} nodes visited for distance {distance}: {route}')
    return route_matrix, distance


def optimize(
    coordinates: CoordinatesVector,
    max_distance: int = None,
    starting_node: int = None,
    closed: bool = True,
    use_nearest_neighbors: bool = False
) -> Tuple[Matrix, int]:
    """Attepmts to find the optimal soultion to a Traveling Salesman problem given a distance matrix and an optional cost-constraint.
    :param coordinates: List of coordinates to use for each node.
    :type coordinates: CoordinatesVector
    :param max_distance: The maximum distance of the solution which attempts to maximize the number of visited nodes, defaults to None
    :type max_distance: int, optional
    :param starting_node: The node to start from as the solution. Only for closed loops if to improve time to converge, defaults to None
    :type starting_node: int, optional
    :param closed: [description], defaults to True
    :type closed: bool, optional
    :param use_nearest_neighbors: Use the nearest_neighbors to solve the problem, instead of branch_and_cut, if possible, defaults to False
    :type use_nearest_neighbors: bool, optional
    :return: The route of the solution in the form of a list of each edge used, along with the distance of the solution.
    :rtype: Tuple[Matrix, int]
    """

    distance_matrix = compute_euclidean_distance_matrix(coordinates)

    if starting_node is not None:
        assert 0 <= starting_node < len(distance_matrix), \
            f"Starting node ({starting_node}) must be within range 0 to {len(distance_matrix)}"

    # use nearest neighbors algorithm
    if (max_distance is not None) or (use_nearest_neighbors is True):

        if starting_node:
            logging.info(f'Starting at node {starting_node}. Algorithm will not iterate to find most optimal solution.')
            route_matrix, distance = nearest_neighbor_path(distance_matrix, max_distance=max_distance, start=starting_node, closed=closed)

        else:
            most_nodes = 0
            route_matrix = None
            distance = None

            for vertex in range(len(distance_matrix)):
                _route_matrix, _distance = nearest_neighbor_path(distance_matrix, max_distance=max_distance, start=vertex, closed=closed)
                logging.info(f'Solution: {np.sum(_route_matrix)} for {distance} from start {vertex}')
                n_nodes = np.sum(_route_matrix)

                if max_distance is not None:  # optimize for most number of nodes
                    # TODO: allow for returning multiple solutions if there is more than one.
                    if n_nodes >= most_nodes:
                        most_nodes = n_nodes
                        route_matrix = _route_matrix
                        distance = _distance

                else:  # optimize for shortest path
                    if distance < _distance:
                        distance = _distance
                        route_matrix = _route_matrix

    else:  # use branch and cut
        if closed is False:
            logging.info('Closed loop argument is False but no max distance is specified. Running branch and cut.')

        route_matrix, distance = branch_and_cut(distance_matrix)

    route = get_edges_from_route_matrix(route_matrix)
    return route, distance


src/tsp_hiram/nearest_nieghbor.py
import random
import math
import logging

logging.basicConfig(level=logging.INFO)

N = ['Antwerp', 'Bruges', 'C-Mine', 'Dinant', 'Ghent',
     'Grand-Place de Bruxelles', 'Hasselt', 'Leuven',
     'Mechelen', 'Mons', 'Montagne de Bueren', 'Namur',
     'Remouchamps', 'Waterloo']

n, V = len(N), set(range(len(N)))

# distances in an upper triangular matrix
dists = [[83, 81, 113, 52, 42, 73, 44, 23, 91, 105, 90, 124, 57],
         [161, 160, 39, 89, 151, 110, 90, 99, 177, 143, 193, 100],
         [90, 125, 82, 13, 57, 71, 123, 38, 72, 59, 82],
         [123, 77, 81, 71, 91, 72, 64, 24, 62, 63],
         [51, 114, 72, 54, 69, 139, 105, 155, 62],
         [70, 25, 22, 52, 90, 56, 105, 16],
         [45, 61, 111, 36, 61, 57, 70],
         [23, 71, 67, 48, 85, 29],
         [74, 89, 69, 107, 36],
         [117, 65, 125, 43],
         [54, 22, 84],
         [60, 44],
         [97],
         []]


# distances matrix
distance_matrix = [[0 if i == j
                    else dists[i][j-i-1] if j > i
                    else dists[j][i-j-1]
                    for j in V] for i in V]


def get_shortest_edge(i, visited, distance_matrix):
    return min([distance_matrix[i][j] for j in V if visited[i][j] == False])

max_distance = 90

def nearest_neighbor_path(distance_matrix, start: int = None, max_distance: int = None) -> List:
    """[summary]

    :param distance_matrix: Simple nearest neighbor algorithm for finding a feasable path for the Traveling Salesman Problem.
    :type distance_matrix: List
    :param start: The node to start to search for a solution, defaults to a randomly chosen node
    :type start: int, optional
    :param max_distance: The maximium distance of the path. If not specific, will find a full path. Otherwise, it will constrain the total cost of the path to be less than or equal to max_distance, defaults to None
    :type max_distance: int, optional
    :return: A feasable solution for the path satisfying the parameters provided.
    :rtype: List
    """


    if start == None:
        start = random.randrange(0, len(distance_matrix))
        logging.info(f'Choosing random starting node: {start}.')

    if max_distance == None:
        max_distance = math.inf

    i = start
    route = [i]

    # Initialize all vertices as unvisited except for self
    visited = [False for v in range(len(distance_matrix))]
    visited[start] = True
    nodes_visited = 1

    distance = 0
    while distance <= max_distance:

        # Have we visited all the nodes?
        if sum(visited) == len(visited):
            distance += distance_matrix[i][start]
            break

        else:
            # Find out the shortest edge connecting the current vertex u and an unvisited vertex v
            shortest_edge = min([distance_matrix[i][j] for j in V if visited[j] == False])

            # can we get home?
            go_home_cost = distance_matrix[i][start]
            if distance + go_home_cost + shortest_edge >= max_distance:
                logging.info(f'lets go home: {distance} + {go_home_cost}')
                distance += go_home_cost
                break

            else:
                distance += shortest_edge
                nearest = [j for j in V if distance_matrix[i][j] == shortest_edge]
                j = nearest[0]

                if len(nearest) > 1:
                    j = random.choice(nearest)
                    logging.info(f'Warning: more than one nearest neighbor for ({i},{j}). Choosing randomly.')

                nodes_visited += 1
                visited[j] = True
                route.append(j)
                i = j

    logging.info(f'{len(route)} nodes visited for distance {distance}: {route}')
    return route


for vertex in V:
    route = nearest_neighbor_path(distance_matrix, start=vertex)